### PR TITLE
Update Intake API spec for v2

### PR DESCRIPTION
**Pull request by simitt**
implements #1276 

# Changes Intake API Json specs for v2 and distributed tracing:
### JSON spec V2
- remove pattern of `error.id`
- remove `transaction.id`
- add `transaction_id` referring to transaction
- add `parent_id`
- add `trace_id`

### ES schema
- add `error.parent_id`
- add `error.trace_id`

### JSON spec
- remove pattern of `transaction.id`
- remove `spans` (don't forbid them, just don't process them)
- add `parent_id`
- add `trace_id` (required attribute)

### ES schema
- add `transaction.parent_id`
- add `transaction.trace_id`

### JSON spec
- change pattern of `id` from long to string
- remove pattern of `transaction_id`
- remove `parent` referring to long id
- add `parent_id` referring to transaction or span
- add `timestamp`
- add `trace_id` (required attribute)

### ES schema
- add `span.parent_id`
- add `span.trace_id`
- add `span.hex_id`
- deprecate `span.id`
- deprecate `span.parent`

## Open Questions:
* Any changes for required optional fields necessary? (@elastic/apm-agent-devs )

**UPDATE**:
After some discussion with @roncohen we decided to also add distributed tracing support to this PR. 
Thus, `parent_id` and `trace_id` have been added.

* The `span.id` is not used for anything, thus will simply be deprecated. (cc @elastic/apm-ui )

### CODE REVIEW

File: docs/spec/errors/v2_error.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{
+    "$id": "docs/spec/errors/v1_error.json",
```
Comment (graphaelli) [inline docs/spec/errors/v2_error.json]: v2_error

File: model/span/_meta/fields.yml [hunk 2]
```diff
@@ -12,9 +12,14 @@
...
-          type: long
+          type: long 
+          description: >
+            Deprecated (>= 6.5). A locally unique ID of the span.
+
+        - name: hex_id
+          type: keyword 
           description: >
-            A locally unique ID of the span.
+            Th ID of the span.
```
Comment (graphaelli) [inline model/span/_meta/fields.yml]: typo

File: model/span/event.go [hunk 3]
```diff
@@ -52,49 +53,80 @@ func ModelSchema() *jsonschema.Schema {
...
 
-func DecodeSpan(input interface{}, err error) (transform.Transformable, error) {
-	if input == nil || err != nil {
+func V2DecodeEvent(input interface{}, err error) (transform.Transformable, error) {
+	e, raw, err := decodeEvent(input, err)
+	if err != nil {
 		return nil, err
 	}
+	decoder := utility.ManualDecoder{}
+	fmt.Println(raw)
```
Comment (roncohen) [inline model/span/event.go]: :)

File: model/span/event.go [hunk 4]
```diff
@@ -43,7 +44,7 @@ var (
 
 	spanDocType = "span"
 
-	processorSpanEntry = common.MapStr{"name": "transaction", "event": spanDocType}
```
Comment (roncohen) [inline model/span/event.go]: for transactions, this is called `processorTransEntry`, and for errors it `processorEntry`. Let's align them on `processorEntry`?

**Comment by roncohen** [comment]
Great work @simitt !

> deprecate id

for 7.0, I assume we'll deprecate `hex_id` and just use `id`? If so, deprecate `id` now could send the wrong signal.

**Comment by simitt** [comment]
@roncohen I suggest to go with @felixbarny 's  suggestion to use `id` instead of `span.hex_id` which aligns with ECS, and deprecate `span.id`. I assume for 7.0 we will align with ECS as much as possible anyways.

**Comment by roncohen** [comment]
@simitt gotcha 👍

**Comment by felixbarny** [comment]
It probably makes sense to deprecate `transaction.id` in favor of `id` in this PR as well. Btw, in [fields.asciidoc](https://github.com/elastic/apm-server/blob/master/docs/fields.asciidoc#service-fields) the `transaction.id` is listed under "service fields" as opposed to "APM transaction fields".

**Comment by simitt** [comment]
We discussed the `fields.asciidoc` issues a while ago with the conclusion to leave as is for now and rework the fields doc generally. 

The `transaction.id` and the `error.id` are currently used and referenced at some places in the dashboards and the UI. The `span.id` isn't afaik. I suggest to keep the ES changes as minimal as possible as this will go out with a minor release. We can aim to align the schema in 7.0

**Comment by felixbarny** [comment]
related: https://github.com/elastic/ecs/issues/99#issuecomment-415029185

> The id you linked from the apm use case has become `event.id`: https://github.com/elastic/ecs#-event-fields

**Comment by simitt** [comment]
I don't think that we should introduce ECS compatibility for the id then with this PR. We would end up with: 
```
{
    "span": { "id": 123, "type": "db", ... },
    "event": { "id": "0123456789abcdef" }, 
    ... 
}
```
I find that very confusing, so I stick with the original idea of introducing a `span.hex_id`.

**Comment by felixbarny** [comment]
Yes, let's make that a different concern. We probably need to discuss what the appropriate ECS name should be first.

**Comment by simitt** [comment]
@sqren could you please check and lmk if the current APM UI is using `span.id` somewhere.

**Comment by sqren** [comment]
@simitt `span.id` is not used anywhere. I thought we did, but it is actually overwritten:

https://github.com/elastic/kibana/blob/02a56b39b10db48b89966b64c1014c16644d9421/x-pack/plugins/apm/server/lib/transactions/spans/get_spans.js#L58

File: CHANGELOG.asciidoc [hunk 5]
```diff
@@ -14,6 +14,8 @@ https://github.com/elastic/apm-server/compare/6.4\...master[View commits]
 
 - Provide basic server information at `/` {pull}1197[1197]
 - Add pipeline registration and pipeline usage {pull}1258[1258]
+- Update JSON schema spec for v2 {pull}1303[1303]
+- Make JSON schema spec distributed tracing compliatn {pull}1303[1303]
```
Comment (roncohen) [inline CHANGELOG.asciidoc]: compliant

File: model/span/event.go (hunk 4 above)
Comment (roncohen) [reply model/span/event.go]: just wanted to make sure you didn't miss this by mistake. It's not critical, so it's fine if you just decided to defer the discussion.

Comment (simitt) [reply model/span/event.go]: I updated now to `processorEntry`.

**Comment by simitt** [comment]
jenkins, retest this please

**Comment by simitt** [comment]
jenkins, retest this please.

**Comment by simitt** [comment]
jenkins, test this please.

**Pull request merged**